Cookbook/Proposals/Danger-Integration.md

```diff
@@ -0,0 +1,152 @@
+
+For example we can write a rule which would post a comment if, inspecting the PR metadata, we see that the `CHANGELOG.md` file was not modified:
+
+![Danger comment](https://danger.systems/images/danger-screenshot-074f084c.png)
+
+This means that:
+
+ - Danger will only trigger when the CI triggers. If we have rules that use the PR metadata to look at what labels are set on the PR for example, they will only consider the labels present at the time the CI job is run, and if those labels get updated, Danger will only see the new labels if the CI job is run again
+ - Danger is more focused on rules that make use either of PR metadata (e.g. checking the title contains a JIRA ticket) or diff content (which files were changed, etc). It does not substitutes to a linter for example (but it can drive a linter like `swiftlint` and use the warnings generated by such linter to generate inline comments in the PR, to report those warnings inline)
+
+Also note that if Danger is run multiple times on a PR – typically because the CI is triggered again after the author added more commits to the PR – it will *update* its comment if it already exists, to prevent spamming the PR with a new comment on every CI build.
+
+
+### Integration of Danger in our CI
+
+We propose to run the `danger` command line as a job:
+
+* Only during the `test_pr` and `test_release` workflows, as it doesn't make sense to run Danger if not in a PR context, but is still useful to have rules checking PRs from both feature and release branches
+* Immediately after the `checkout_code` job – so that danger will have access to the code itself if needed – and in parallel with all other jobs (`test_*` and `build_*` jobs) in the workflows
+
+⏱ Danger is quite fast, as all the information the rules are based on are the metadata fetched from the Pull Request or from the git diff, so it doesn't have to analyse your codebase (like a linter would do) – just to decode the JSON of the PR, run the `Dangerfile`, then call the GitHub API to post the PR comment. It's unlikely that integrating Danger will have any impact on the CI total time, especially if run in parallel of the `test_pr` jobs
+
+
+### Initial Rule to activate
+
+_As this doesn't make sense to integrate Danger in the repo without having any rule at all in the `Dangerfile` (as this would do nothing and thus would not solve any issue listed in the Motivation section), this proposal also suggest to implement one simple but useful rule during this initial Danger integration in our repo._
+
+The first rule that we propose to activate, and one which can be easy to write, is one to warn about a Pull Request which exceeds the maximum number of changes
+
+This simply consists of this code in the `Dangerfile`:
+
+```ruby
+warn("This PR is too big. Please split it in smaller PRs. If you can't split it, please justify why.") if git.lines_of_code > 800
```

Comment:
@ilyapuchka That's a great question.

The short answer is that Swift Danger is sure easier to use to write Dangerfiles, but way less advanced in terms of possibilities and plugin support, so I see a risk of limiting us when writing the rules. Orta even says on the "[Help me choose](https://danger.systems/js/js-vs-ruby.html)" page:

> You may notice that there was a Swift logo in the index page, Swift is the first language supported by Danger JS’s process environment. Once a new language has been built out, and is stable, then I’ll be building out some documentation in this site.

(For example, plugins for `swliftint` or similar are not available in Danger Swift yet afaik)

* [ ] I'll add that to "Alternatives Considered" though 👍 

As for the details of rule implementation, it's also mentioned already in the "Alternatives Considered" section.

* [ ] I'll add the idea about ignoring localisation or snapshots, great idea 👍

Reply:
true, but then again I doubt there is anything tricky in ruby plugins that would block rewriting them in Swift, unless the Danger framework for Swift does not have a feature parity with Danger module for Ruby

Reply:
It does not 100% sadly (well things might have been better since last time I looked, but some PR metadata are not retranscripted in the Swift DSL for example, requiring to parse the raw metadata ourselves and defeating the benefits of the DSL)